File: src/code/java/_020.java
package code.java;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class _020 {
    Map<Character, Character> map;

    public _020() {
        map = new HashMap<Character, Character>();
        map.put('}', '{');
        map.put(']', '[');
        map.put(')', '(');
    }

    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        int len = s.length();
        for (int i = 0; i < len; i++) {
            // if 左括号 放进stack
            // if 右括号 能否消除 ? 消除 : 返回false
            char c = s.charAt(i);
            if ("{[(".indexOf(c) != -1) {
                stack.push(c);
            } else if ("}])".indexOf(c) != -1) {
                if (stack.isEmpty() || map.get(c) != stack.pop()) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}

File: src/code/js/_020.js
/**
 * @param {string} s
 * @return {boolean}
 */
var map = {"}": "{", "]": "[", ")": "("};
var isValid = function (s) {
    var stack = [];
    for (var i = 0; i < s.length; i++) {
        //if 左括号 放进stack
        //if 右括号 能否消除 ? 消除 : 返回false
        var c = s[i];
        if ("{[(".indexOf(c) > -1) {
            stack.push(c);
        } else if ("}])".indexOf(c) > -1) {
            if (stack.pop() !== map[c]) {
                return false;
            }
        }
    }
    return stack.length === 0;
};